Firestore/core/src/nanopb/nanopb_util.h

```diff
@@ -104,6 +104,69 @@ T* _Nonnull MakeArray(pb_size_t count) {
   return static_cast<T*>(calloc(count, sizeof(T)));
 }
 
+template <typename T>
+T* _Nonnull ResizeArray(void* _Nullable ptr, pb_size_t count) {
+  return static_cast<T*>(realloc(ptr, count * sizeof(T)));
+}
+
+/**
+ * Initializes a repeated field with a list of values. Applies `converter` to
+ * each value before assigning.
+ */
+template <typename T, typename Iterator, typename Func>
+void SetRepeatedField(T** fields_array,
```

Comment:
Optional: returning `absl::span` could allow you to get rid of the output parameters. Consider something like:
```cpp
template <typename Iterator, typename Func>
auto SetRepeatedField(Iterator first, Iterator last, const Func& converter)
    -> absl::span<decltype(converter())> {
  using T = decltype(converter());

  auto size = CheckedSize(std::distance(first, last));
  absl::span<T> result{MakeArray<T>(size), size};
  for (auto* e : result) {
    *e = converter(*first);
    ++first;
  }

  return result;
}
```

Reply:
I think that's quite a lot, actually. Additionally, it adds the possibility of passing null arguments, which imposes the extra burden of checking the arguments and finding a way to signal error. Also, see go/cstyle#Inputs_and_Outputs:
> Prefer using return values over output parameters: they improve readability, and often provide the same or better performance

Reply:
I assume you meant to say "I don't think that's quite a lot, actually."

Changing this would negatively impact all callsites. It would change code such as this https://github.com/firebase/firebase-ios-sdk/pull/8008/files#diff-4446e45495c7ddead4c8e19f3ad9233bd1ba6f0eac229187c20d66d4165f0850R433  from:


``` 
google_firestore_v1_ArrayValue array_value;
SetRepeatedField(&array_value.values, &array_value.values_count, operands);
```

to

``` 
google_firestore_v1_ArrayValue array_value;
auto span = SetRepeatedField(&array_value.values, &array_value.values_count, operands); 
array_value.values = span.begin();
array_value.values_count = static_cast<pb_size_t>(span.size());
```

It creates an extra object and introduces the possibility of not assigning all values, which is one of the main reasons why I created this helper.

Reply:
Thanks for the example! I didn't realize the array was nested inside another message (though I guess it should have been obvious).

This could work using `std::tie` and `std::pair` (unfortunately, `span` wouldn't work) to emulate multiple assignment:
```cpp
google_firestore_v1_ArrayValue array_value;
std::tie(array_value.values, array_value.values_count) = SetRepeatedField(operands);
```

If you don't want to deal with `std::tie`, it's absolutely fine to leave as is (but please check the input pointer arguments for null).

Reply:
Null checks added.